src/corehost/cli/fxr/sdk_resolver.cpp

```diff
@@ -5,183 +5,240 @@
-    catch (const std::exception& je)
+
+    if (requested.empty() || !sdk_info::print_all_sdks(dotnet_root, _X("  ")))
+    {
+        trace::error(_X("  It was not possible to find any installed .NET Core SDKs"));
+        trace::error(_X("  Did you mean to run .NET Core SDK commands? Install a .NET Core SDK from:"));
+        trace::error(_X("      %s"), DOTNET_CORE_DOWNLOAD_URL);
+    }
+    return {};
+}
+
+sdk_resolver sdk_resolver::from_nearest_global_file(bool allow_prerelease)
+{
+    pal::string_t cwd;
+    if (!pal::getcwd(&cwd))
+    {
+        trace::verbose(_X("Failed to obtain current working dir"));
+        assert(cwd.empty());
+    }
+    else
     {
-        pal::string_t jes;
-        (void)pal::utf8_palstring(je.what(), &jes);
-        trace::error(_X("A JSON parsing exception occurred in [%s]: %s"), global_json.c_str(), jes.c_str());
+        trace::verbose(_X("--- Resolving .NET Core SDK with working dir [%s]"), cwd.c_str());
     }
-    trace::verbose(_X("CLI version is [%s] in global json file [%s]"), retval.c_str(), global_json.c_str());
-    return retval;
+    return from_nearest_global_file(cwd, allow_prerelease);
 }
 
-pal::string_t resolve_sdk_version(pal::string_t sdk_path, bool disallow_prerelease, pal::string_t global_cli_version)
+sdk_resolver sdk_resolver::from_nearest_global_file(const pal::string_t& cwd, bool allow_prerelease)
 {
-    fx_ver_t specified;
+    sdk_resolver resolver{ allow_prerelease };
 
-    //   Validate the global cli version if specified
-    if (!global_cli_version.empty())
+    if (!resolver.parse_global_file(find_nearest_global_file(cwd)))
     {
-        if (!fx_ver_t::parse(global_cli_version, &specified, false))
-        {
-            trace::error(_X("The specified SDK version '%s' could not be parsed"), global_cli_version.c_str());
-            return pal::string_t();
-        }
+        // Fall back to a default SDK resolver
+        resolver = sdk_resolver{ allow_prerelease };
 
-        // Always consider prereleases when the version specified in global.json is itself a prerelease
-        if (specified.is_prerelease())
-        {
-            disallow_prerelease = false;
-        }
+        trace::error(
+            _X("Ignoring SDK settings in global.json: the latest installed .NET Core SDK (%s prereleases) will be used"),
+            resolver.allow_prerelease() ? _X("including") : _X("excluding"));
     }
 
-    trace::verbose(_X("--- Resolving SDK version from SDK dir [%s]"), sdk_path.c_str());
+    // If the requested version is a prerelease, always allow prerelease versions
+    if (resolver._requested.is_prerelease())
+    {
+        resolver._allow_prerelease = true;
+    }
 
-    pal::string_t retval;
-    std::vector<pal::string_t> versions;
+    return resolver;
+}
 
-    pal::readdir_onlydirectories(sdk_path, &versions);
-    fx_ver_t max_ver;
-    for (const auto& version : versions)
+sdk_roll_forward_policy sdk_resolver::to_policy(const pal::string_t& name)
```

Comment:
This member needs to support case-insensitive names based on a discussion from the design document.